recipes/AceEditorHalogenHooks/src/Main.purs

```diff
@@ -0,0 +1,100 @@
+module AceEditorHalogenHooks.Main where
+
+import Prelude
+
+import Ace as Ace
+import Ace.EditSession as Session
+import Ace.Editor as Editor
+import Data.Foldable (traverse_)
+import Data.Maybe (Maybe(..))
+import Data.Symbol (SProxy(..))
+import Data.Tuple.Nested ((/\))
+import Effect (Effect)
+import Effect.Aff.Class (class MonadAff)
+import Halogen (liftEffect)
+import Halogen as H
+import Halogen.Aff as HA
+import Halogen.HTML as HH
+import Halogen.HTML.Events as HE
+import Halogen.HTML.Properties as HP
+import Halogen.Hooks as Hooks
+import Halogen.Query.EventSource as ES
+import Halogen.VDom.Driver (runUI)
+
+main :: Effect Unit
+main =
+  HA.runHalogenAff do
+    body <- HA.awaitBody
+    void $ runUI containerComponent unit body
+
+_ace = SProxy :: SProxy "ace"
+
+containerComponent
+  :: forall unusedQuery unusedInput unusedOutput anyMonad
+   . MonadAff anyMonad
+  => H.Component HH.HTML unusedQuery unusedInput unusedOutput anyMonad
+containerComponent = Hooks.component \rec _ -> Hooks.do
+  msg /\ msgIdx <- Hooks.useState ""
+  Hooks.pure $
+    HH.div_
+      [ HH.h1_
+          [ HH.text "ace editor" ]
+      , HH.div_
+          [ HH.p_
+              [ HH.button
+                  [ HE.onClick \_ -> Just do
+                      void $ Hooks.query rec.slotToken _ace unit $ H.tell (ChangeText "")
+                  ]
+                  [ HH.text "Clear" ]
+              ]
+          ]
+      , HH.div_
+          [ HH.slot _ace unit aceComponent unit \(TextChanged t) -> Just do
+              Hooks.put msgIdx t
+          ]
+      , HH.p_
+          [ HH.text ("Current text: " <> msg) ]
+      ]
+
+data ParentAction
+  = ClearText
+  | HandleAceUpdate AceOutput
+
+data AceQuery a = ChangeText String a
+
+data AceOutput = TextChanged String
+
+aceComponent
+  :: forall unusedInput anyMonad
+   . MonadAff anyMonad
+  => H.Component HH.HTML AceQuery unusedInput AceOutput anyMonad
+aceComponent = Hooks.component \rec _ -> Hooks.do
+  state /\ stateIdx <- Hooks.useState Nothing
```

Comment:
It's nice to use annotations when the type is not obvious from the initial value.
For example:
```purs
Hooks.useState (Nothing :: Maybe Editor)
```
Although I'm thinking `Maybe Document` would be better (as mentioned earlier).
Also, should `state` be renamed to something more specific? Downside is that requires some de-shadowing with duplicated names.

Reply:
I did use clarify the `Nothing` type [in this commit](https://github.com/JordanMartinez/purescript-cookbook/pull/162/commits/74295e720a7557faa0a5405581cd66d7562023b5#diff-8a44b4282b62e6e07657f75e3dba6ef7R76)

Not sure about the `state` renaming...